sklearn/utils/testing.py

```diff
@@ -372,14 +375,52 @@ def __exit__(self, *exc_info):
     assert_greater = _assert_greater
 
 
+def _sparse_dense_allclose(val1, val2, rtol=1e-7, atol=0):
```

Comment:
The question is: would an `assert_safe_sparse_allclose` actually allow some refactoring of our current tests?

Yes!! From a quick check I found a few places where the sparse matrices were densified before checking!